pyoverkiz/models.py

```diff
@@ -181,7 +188,19 @@ def __init__(
         self.data_properties = data_properties
         self.type = ProductType(type)
         self.place_oid = place_oid
-        self.protocol = Protocol(self.device_url.split(":")[0])
+
+        match = re.search(DEVICE_URL_RE, device_url)
+
+        self.subsystem_id = None
+        self.is_sub_device = False
+        if match:
+            self.protocol = Protocol(match.group("protocol"))
+            self.gateway_id = match.group("gatewayId")
+            self.device_address = match.group("deviceAddress")
```

Comment:
We check for `if match`, but these variables are not optional in the device object. How do we handle this when we don't have a match? (can this happen?). Do we always have matches in all our test files?

Reply:
I think that's the case.  
In test_client, a Setup is created with all json files.  
On init of Setup, we have   
`self.devices = [Device(**d) for d in devices]`
So I guess yes, all elements from test are correct.

I also reviewed all "deviceURL" in tests file adn all are formed corresponding to the regex.

Reply:
Ok, but this `if match` should not be necessary in that case? If there is no match, self.gateway_id and self.device_address won't be set.

Reply:
See here, because of mypy...
https://github.com/iMicknl/python-overkiz-api/pull/639#pullrequestreview-1145241149

Reply:
I don't see what the relation is with this `if match`?

Reply:
Without it, mypy tells me:  
`Item "None" of "Optional[Match[str]]" has no attribute "group"`

Reply:
What's surprise me is that mypy does not complain about missing gateway id if there is no match...

Reply:
I guess it's because of :  
https://docs.python.org/3/library/re.html#re.Match.group

> If the regular expression uses the (?P<name>...) syntax, the groupN arguments may also be strings identifying groups by their group name. If a string argument is not used as a group name in the pattern, an [IndexError](https://docs.python.org/3/library/exceptions.html#IndexError) exception is raised.

No match found, exception is raised.